telemetry/client.go

```diff
@@ -10,70 +10,207 @@ import (
-	v1protocol "github.com/status-im/status-go/protocol/v1"
+	"github.com/status-im/status-go/wakuv2"
 
 	v2protocol "github.com/waku-org/go-waku/waku/v2/protocol"
 )
 
+type TelemetryType string
+
+const (
+	ProtocolStatsMetric    TelemetryType = "ProtocolStats"
+	ReceivedEnvelopeMetric TelemetryType = "ReceivedEnvelope"
+	SentEnvelopeMetric     TelemetryType = "SentEnvelope"
+	UpdateEnvelopeMetric   TelemetryType = "UpdateEnvelope"
+	ReceivedMessagesMetric TelemetryType = "ReceivedMessages"
+)
+
+type TelemetryRequest struct {
+	Id            int              `json:"id"`
+	TelemetryType TelemetryType    `json:"telemetry_type"`
+	TelemetryData *json.RawMessage `json:"telemetry_data"`
+}
+
+func (c *Client) PushReceivedMessages(receivedMessages wakuv2.ReceivedMessages) {
+	c.receivedMessagesCh <- receivedMessages
+}
+
+func (c *Client) PushSentEnvelope(sentEnvelope wakuv2.SentEnvelope) {
+	c.sentEnvelopeCh <- sentEnvelope
+}
+
+func (c *Client) PushReceivedEnvelope(receivedEnvelope *v2protocol.Envelope) {
+	c.receivedEnvelopeCh <- receivedEnvelope
+}
+
 type Client struct {
-	serverURL  string
-	httpClient *http.Client
-	logger     *zap.Logger
-	keyUID     string
-	nodeName   string
+	serverURL          string
+	httpClient         *http.Client
+	logger             *zap.Logger
+	keyUID             string
+	nodeName           string
+	version            string
+	receivedMessagesCh chan wakuv2.ReceivedMessages
+	receivedEnvelopeCh chan *v2protocol.Envelope
+	sentEnvelopeCh     chan wakuv2.SentEnvelope
+	telemetryCh        chan TelemetryRequest
+	nextId             int
+	sendPeriod         time.Duration
 }
 
-func NewClient(logger *zap.Logger, serverURL string, keyUID string, nodeName string) *Client {
+func NewClient(logger *zap.Logger, serverURL string, keyUID string, nodeName string, version string) *Client {
 	return &Client{
-		serverURL:  serverURL,
-		httpClient: &http.Client{Timeout: time.Minute},
-		logger:     logger,
-		keyUID:     keyUID,
-		nodeName:   nodeName,
+		serverURL:          serverURL,
+		httpClient:         &http.Client{Timeout: time.Minute},
+		logger:             logger,
+		keyUID:             keyUID,
+		nodeName:           nodeName,
+		version:            version,
+		receivedMessagesCh: make(chan wakuv2.ReceivedMessages),
+		receivedEnvelopeCh: make(chan *v2protocol.Envelope),
+		sentEnvelopeCh:     make(chan wakuv2.SentEnvelope),
+		telemetryCh:        make(chan TelemetryRequest),
+		nextId:             0,
+		sendPeriod:         10 * time.Second,
+	}
+}
+
+func (c *Client) CollectAndProcessTelemetry() {
+	go func() {
+		for {
+			select {
+			case receivedMessages := <-c.receivedMessagesCh:
+				c.processAndPushTelemetry(receivedMessages)
+			case receivedEnvelope := <-c.receivedEnvelopeCh:
+				c.processAndPushTelemetry(receivedEnvelope)
+			case sentEnvelope := <-c.sentEnvelopeCh:
+				c.processAndPushTelemetry(sentEnvelope)
+			}
+		}
+	}()
+}
+
+func (c *Client) Start() {
```

Comment:
Should we pass a context to the Start() function?
We could use it for killing the go routine once logout happens, i.e.:

```
select {
...
case <-ctx.Done():
    return
...
}
```